Add YANG parser

xray7224 opened the pull request:
Add a yang parser from the YANG data structure to lua tables.

wingo commented on src/lib/yang/parser.lua:
```diff
@@ -0,0 +1,341 @@
+local Leaf = new_type()
+function Leaf.new(properties)
+   return setmetatable(properties, Leaf)
+end
```
Probably can remove new_type() and Leaf I think since they are unused; just inline new_type() into its one use below defining Parser

wingo commented on src/lib/yang/parser.lua:
```diff
@@ -0,0 +1,341 @@
+module(..., package.seeall)
```
Can we add an apache header like this please:

```lua
-- Use of this source code is governed by the Apache 2.0 license; see COPYING.

```

wingo commented on src/lib/yang/parser.lua:
```diff
@@ -0,0 +1,341 @@
+   test_string(lines("   'foo",
+		     " bar'"),
+	       lines("foo", "bar"))
+
```
just one line of vertical space please

wingo commented:
Looking great.  Can you please add test cases for all of the grammar productions?  I guess just "simple statement without argument", "simple statement with argument", "compound statement with argument", "compount statement without argument".  Where compound statement is of the form "foo {...}".  I know that parsing the example file is pretty much good enough, but still it's good to have known expectations for every bit.

Probably we are also missing a comment at the top indicating how the parser represents its result.  But, maybe we can just comment the one entry function, because really the module should probably just export one function to parse a module, because that's the only thing we need to do; wdyt?

xray7224 commented:
I've added the test cases which were missing that you asked for, it's a good job I did as there was actually a bug in the parsing of a simple statement without an argument. I fixed that bug. I also added some documentation to the top explaining what the module provides and what should be expected. I decided to add it at the top as you first suggested because I figured that'd probably be the first place someone would look if unsure, it explains which functions/methods should be called and what one should expect as a return value.

If you could re-review that would be great!

wingo commented on src/lib/yang/parser.lua:
```diff
@@ -0,0 +1,365 @@
+	       lines("foo", "    bar"))
+   test_string(lines("   'foo", " bar'"),
+	       lines("foo", "bar"))
+
```
just one nit here, can you have just one line of whitespace between "blocks" and not two, please; thanks :)  same below.

wingo commented on src/lib/yang/parser.lua:
```diff
@@ -0,0 +1,365 @@
+   local fin = assert(io.open("example.yang"))
+   local yangexample = fin:read("*a")
+   local parser = Parser.new(yangexample, "example.yang")
+   parser:parse_module()
```
would you mind adding two top-level functions, `function parse_file(filename)` and `function parse_string(str)`, which do this dance of instantiating a parser etc?  Then you could replace the manual instantiation / `parse_module` method invocations here with a call to parse_string or parse_module.  Sorry for the nitpicking but it's always nice to stop work at a good point, ux-wise :)

wingo commented:
Great catch!  Just one syntax nit and a little function addition to do.

xray7224 commented:
Fixed those two problems you listed.

wingo commented:
LGTM; thank you for bearing with all the nits!
